### PR TITLE
[serialization] Reject dunder attributes in tensor state under weights_only

### LINKED ISSUE #195607: `_set_obj_state` applies unrestricted setattr from checkpoint state under weights_only=True
## Summary

When loading a checkpoint with `torch.load(..., weights_only=True)`, `_set_obj_state()` in `torch/_utils.py` calls `setattr(obj, k, v)` using keys taken straight from the pickle stream, with no filtering. A crafted state dict can set dunder attributes like `__class__` on a loaded object, which silently changes its type (for example `nn.Parameter` into a plain `Tensor`) with no error or warning.

To be clear, this is not a security vulnerability. `weights_only=True` guarantees safe deserialization (no RCE, no out-of-bounds writes), and the values here still go through the restricted unpickler. This is a hardening request. A normal `torch.save` never emits these keys, so rejecting them is cheap and closes off a class of surprising, hard-to-debug behavior from malformed or hand-crafted checkpoints.

## Location

- File: `torch/_utils.py`, function `_set_obj_state()`
- Callers: `_rebuild_parameter_with_state` and `torch._tensor._rebuild_from_type_v2`, both on the restricted unpickler's allowlist

```python
def _set_obj_state(obj, state):
    ...
    for k, v in dict_state.items():
        setattr(obj, k, v)   # k comes from the pickle stream, unfiltered
```

## What happens

The restricted unpickler validates a lot of operations. `GLOBAL` is checked against an allowlist, `REDUCE` only permits allowed functions, and `SETITEM`/`SETITEMS` is restricted to `dict`, `OrderedDict`, and `Counter`. What it doesn't restrict is which attribute names can be set on objects returned by allowed functions. `_set_obj_state` is the one path where a key from the stream reaches `setattr` directly.

Since `Parameter` inherits directly from `Tensor` (same C-level layout), CPython allows a `__class__` reassignment, so a state dict of `{"__class__": torch.Tensor}` downgrades a `Parameter` to a plain `Tensor`. `isinstance(x, nn.Parameter)` then returns `False`, which quietly breaks optimizers, gradient tracking, and model-surgery tools downstream, with no signal at load time.

The same mechanism also lets you inject arbitrary attributes (for example `{"_admin": True}`) into `__dict__`, or set `__reduce__` or `__setattr__` to non-callable values, which then raises `TypeError` later on re-serialization or attribute assignment.

## Reproduction

Reproduced on PyTorch 2.13.0+cu130, Python 3.13, Colab (Tesla T4). Self-contained, no downloads needed.

```python
import struct, tempfile, zipfile, os
import torch

PROTO=b'\x80'; STOP=b'.'; GLOBAL=b'c'; REDUCE=b'R'
MARK=b'('; TUPLE=b't'; EMPTY_DICT=b'}'; EMPTY_TUPLE=b')'
SETITEM=b's'; SETITEMS=b'u'; BINPUT=b'q'
BINUNICODE=b'X'; NEWFALSE=b'\x89'; NEWTRUE=b'\x88'
BININT1=b'K'; BINPERSID=b'Q'

def ustr(s):
    b = s.encode('utf-8')
    return BINUNICODE + struct.pack('<I', len(b)) + b

def make_storage():
    return (MARK + ustr("storage") + GLOBAL + b'torch\nFloatStorage\n' +
            ustr("0") + ustr("cpu") + BININT1 + bytes([6]) + TUPLE + BINPERSID)

def make_tensor():
    return (GLOBAL + b'torch._utils\n_rebuild_tensor_v2\n' + MARK +
            make_storage() + BININT1 + bytes([0]) +
            MARK + BININT1 + bytes([3]) + BININT1 + bytes([2]) + TUPLE +
            MARK + BININT1 + bytes([2]) + BININT1 + bytes([1]) + TUPLE +
            NEWFALSE + GLOBAL + b'collections\nOrderedDict\n' + EMPTY_TUPLE +
            REDUCE + TUPLE + REDUCE)

pkl = (PROTO + bytes([2]) +
       GLOBAL + b'collections\nOrderedDict\n' + EMPTY_TUPLE + REDUCE +
       BINPUT + bytes([0]) + MARK + ustr('param') +
       GLOBAL + b'torch._utils\n_rebuild_parameter_with_state\n' + MARK +
       make_tensor() + NEWTRUE +
       GLOBAL + b'collections\nOrderedDict\n' + EMPTY_TUPLE + REDUCE +
       EMPTY_DICT + ustr('__class__') + GLOBAL + b'torch\nTensor\n' +
       SETITEM + TUPLE + REDUCE + BINPUT + bytes([1]) +
       SETITEMS + STOP)

path = tempfile.mktemp(suffix='.pt')
with zipfile.ZipFile(path, 'w') as z:
    z.writestr('archive/data.pkl', pkl)
    z.writestr('archive/version', '3\n')
    z.writestr('archive/byteorder', 'little')
    z.writestr('archive/.format_version', '1')
    z.writestr('archive/.storage_alignment', '64')
    z.writestr('archive/.data/serialization_id', '0' * 40)
    z.writestr('archive/data/0', struct.pack('<6f', 1, 2, 3, 4, 5, 6))

result = torch.load(path, weights_only=True)
param = result['param']
os.unlink(path)

print(type(param).__name__)                   # Tensor  (expected: Parameter)
print(isinstance(param, torch.nn.Parameter))  # False   (expected: True)
```

Observed output:
Tensor
False

Expected: `Parameter` and `True`. The `Parameter` is silently downgraded to a `Tensor` with no error or warning during load.

## Suggested hardening

Reject dunder keys in `_set_obj_state`, since a legitimate `torch.save` never produces them:

```python
for k, v in dict_state.items():
    if isinstance(k, str) and k.startswith('__') and k.endswith('__'):
        raise RuntimeError(
            f"_set_obj_state: setting dunder attribute '{k}' is not allowed"
        )
    setattr(obj, k, v)
```

The same check should go on the `slots_state` loop below it. Happy to open a PR if this seems reasonable.

### Versions

PyTorch version: 2.13.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.10
Libc version: glibc-2.35

Python version: 3.13.15 (main, Aug  6 2026, 11:06:23) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.6.122+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.8.93
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 580.82.07
Is XNNPACK available: False

cc @mruberry @mikaylagawarecki

### PR DESCRIPTION
**Pull request by JHC56**
Fixes #195607

## Summary

`_set_obj_state` restores a checkpoint's Python attributes with `setattr(obj, k, v)`. The keys come straight from the pickle stream, unfiltered. Under `weights_only=True` that's the one place a stream-controlled key reaches `setattr`.

So a crafted state dict can set dunder attributes. `{"__class__": torch.Tensor}` is the interesting one: Parameter and Tensor share a layout, so CPython allows the reassignment. The Parameter quietly becomes a plain Tensor and `isinstance(x, nn.Parameter)` starts returning False. No error, no warning.

This isn't an RCE, and values still go through the restricted unpickler, so it's a hardening change rather than a security fix. A normal `torch.save` never writes dunder keys here. Rejecting them is cheap and avoids a class of confusing failures from malformed or hand-crafted checkpoints.

## Change

`_set_obj_state` is left alone. The check now lives in `_weights_only_unpickler.py` instead.

Two allowlisted rebuild functions hand a stream-controlled state dict to `_set_obj_state`: `_rebuild_parameter_with_state` and `_rebuild_from_type_v2`. Their allowlist entries now point at thin wrappers that reject dunder-named keys. GLOBAL and REDUCE both resolve through that same dict, so the wrappers are always what a restricted load gets, and nothing else is affected.

One detail: `_rebuild_from_type_v2` skips `_set_obj_state` when the subclass has its own `__setstate__`. The wrapper mirrors that, so those subclasses keep full control of their state.

## Test plan

Checked on 2.13.0:

- The crafted checkpoint from the issue now raises `UnpicklingError` instead of quietly handing back a `Tensor`.
- `nn.Parameter` with a custom attribute, a plain `state_dict`, and `TwoTensor` all still round-trip under `weights_only=True`.
- `weights_only=False` still restores dunder-named state that weights-only rejects.
- Subclasses with their own `__setstate__` are unaffected.
- Non-dict `__getstate__` results are ignored.

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
I think this hardening is being applied at a broader boundary than the PR intends.

The new dunder rejection lives directly in `_set_obj_state()`, but that helper is part of ordinary Tensor/Parameter reconstruction as well; it is not specific to the weights-only unpickler. `_get_obj_state()` also deliberately serializes a user-defined `__getstate__()` result, or the object's `__dict__`/slots.

That means this changes normal `torch.save` / `torch.load(weights_only=False)` behaviour too: a Tensor/Parameter subclass whose legitimate Python state contains a dunder-named attribute can be serialized today but will now fail during reconstruction. The PR description says this is a `weights_only=True` hardening change and that valid inputs are unaffected, but the implementation enforces the policy globally.

Could the rejection instead be applied specifically on the restricted weights-only reconstruction path, or be controlled by an explicit restricted-load flag rather than changing `_set_obj_state()` for every caller?

I would also add a BC regression showing that ordinary unrestricted loading can still restore user-defined state that weights-only deliberately rejects.

**Comment by JHC56** [comment]
Thanks for the review, that's a fair point. `_set_obj_state` is shared across both paths, so the original change affected unrestricted loading too.

I've reverted it and scoped the rejection to the weights-only path instead. The allowlist entries for `_rebuild_parameter_with_state` and `_rebuild_from_type_v2` now point at thin wrappers that reject dunder keys. Those are the only two allowlisted rebuild functions that hand a stream-controlled state dict to `_set_obj_state`, and GLOBAL and REDUCE both resolve through the same dict, so no flag is needed and the unrestricted path is untouched.

Also added the BC regression you asked for, plus a case confirming that subclasses with their own `__setstate__` are unaffected, since those never reach `_set_obj_state`.
